### ISSUE TITLE
Bug: incorrect kube metadata

### ISSUE THREAD

**Issue by kushwiz** [fluent/fluent-bit#353]
Fluent-bit version used: `0.11.16`
Configuration used:
```
[SERVICE]
    Daemon off
    Parsers_file parser

[INPUT]
    Name tail
    Path /var/log/containers/*othernamespace*.log
    Parser docker
    Tag otherns.*

[INPUT]
    Name tail
    Path /var/log/containers/*namespace*.log
    Parser docker
    Tag ns.*

[FILTER]
    Name kubernetes
    Match *

[OUTPUT]
    Name stdout
    Match ns.*
```

Logs from pods of `namespace` get appended with Kube metadata of pods from `othernamespace`.   Some logs do not have metadata information as well.  

Let me know if you cannot replicate this.

**Comment by edsiper** [comment]
@kushwiz 

I've tried to replicate without success...

note: logs without metadata usually happens when the API server do not have the meta for the key pair of namespace and pod_name .

**Comment by edsiper** [comment]
are you able to replicate with minikube ?

**Comment by kushwiz** [comment]
I have a k8s cluster running with two namespaces. No not minikube.

**Comment by edsiper** [comment]
is it enough to have two "echo something" PODS on each namespace to replicate the problem ?

**Comment by kushwiz** [comment]
I had enough number of logs - approx 1000 lines from from both namespaces.  I believe its some race condition? I am not sure.

**Comment by edsiper** [comment]
I've tried to replicate again in a real cluster (digital ocean) but without success, I did this:

#### Create namespaces

```bash
$ kubectl create namespace ns1
$ kubectl create namespace ns2
$ kubectl create namespace logging
```

#### Create service account and roles

```bash
$ kubectl create -f https://raw.githubusercontent.com/edsiper/fluent-bit-ds/master/fluent-bit-service-account.yaml 
$ kubectl create -f https://raw.githubusercontent.com/edsiper/fluent-bit-ds/master/fluent-bit-role.yaml
$ kubectl create -f https://raw.githubusercontent.com/edsiper/fluent-bit-ds/master/fluent-bit-role-binding.yaml
```

#### Deploy sample PODS

Each pod runs in a different namespace and generate 1k lines:

```
$ kubectl create -f https://raw.githubusercontent.com/edsiper/fluent-bit-ds/master/pod1.yaml --record
$ kubectl create -f https://raw.githubusercontent.com/edsiper/fluent-bit-ds/master/pod2.yaml --record
```
#### Deploy Fluent Bit DaemonSet

```bash
$ kubectl create -f https://raw.githubusercontent.com/edsiper/fluent-bit-ds/master/fluent-bit-ds.yaml
```
The daemonset will do nothing with looks since this is a test config, so we need to get into one of the fluent-bit container and run it manually:

```
$ kubectl exec -ti fluent-bit-ml0lx --namespace=logging -- sh
```

install and run bash in the container to make things easier:

```
$ apt-get install bash
$ bash
```
run fluent bit from the container bash session:

```
$ cd /fluent-bit
$ bin/fluent-bit -R etc/parsers.conf -i tail -p 'path=/var/log/containers/pod-ns*.log' -p parser=docker -p tag='kube.*' -F kubernetes -m '*' -o stdout -f 1 -vv
```

on my case I saw all logs POD samples properly processed with kubernetes metadata appended. 

If we can get a reproducible case based on the setup provided would be great for troubleshooting.

**Comment by kushwiz** [comment]
Can you try with two in_tail blocks ?

tail for /var/log/containers/*ns1*.log . Tag ns1.*
tail for /var/log/containers/*ns2*.log . Tag ns2.*

Something like below.

[SERVICE]
    Daemon off
    Parsers_file parser

[INPUT]
    Name tail
    Path /var/log/containers/*ns1*.log
    Parser docker
    Tag ns1.*

[INPUT]
    Name tail
    Path /var/log/containers/*ns2*.log
    Parser docker
    Tag ns2.*

[FILTER]
    Name kubernetes
    Match *

[OUTPUT]
    Name stdout
    Match ns1.*

**Comment by edsiper** [comment]
Just tested with:

```
bin/fluent-bit  \ 
    -R etc/parsers.conf \
    -i tail -p 'path=/var/log/containers/pod-ns1*.log' \
             -p tag='ns1.*' \
             -p parser=docker  \
    -i tail -p 'path=/var/log/containers/pod-ns2*.log' \
             -p tag='ns2.*' \
             -p parser=docker \
    -F kubernetes -m '*'  \
    -o stdout -m 'ns1.*' -f 1 -vv
```

**Comment by kushwiz** [comment]
Let me debug this at my end. You are using namespace:pod_name as the cache key. I think in my case, there might have been some collision.

**Comment by kushwiz** [comment]
Yes, i see that the flb_hash_get returns wrong metadata.  We can chat about this on slack. I am in the fluentd slack group.

**Comment by kushwiz** [comment]
At: https://github.com/fluent/fluent-bit/blob/master/src/flb_hash.c#L239

If you have just one entry in the MonkeyList, you should again compare the string and set to NULL if it doesn't match.  Does this make sense? Let me submit a PR.

**Comment by kushwiz** [comment]
https://github.com/fluent/fluent-bit/pull/355 

This fix worked for me. It does make sense to string compare the keys even if there is just one entry in the list?

**Comment by edsiper** [comment]
Fixed by #355 

thanks for troubleshooting and fix this. I am releasing a new version with this change now.

**Comment by edsiper** [comment]
http://fluentbit.io/announcements/v0.11.17/
